### DISCUSSION TITLE
Next/link not following redirects defined in next.config.js

### DISCUSSION THREAD

**Discussion by reckter** [vercel/next.js#26426] [Help] [answered]
# Bug report

When specifying redirects in the next.config.js on next.js 9.5.3, `next/link` on the page will not pick that up that redirect correctly, and not redirect accordingly, leading to a navigation to the source of the redirect (instead of the destination).

I made a minimal repository with said error: https://github.com/homefully/next-js-9-5-3-link-redirect-bug

Steps:
- next.js config with a redirect (here from `/failure` to `/success`)
- `<Link passHref href="/failure">...</Link>` 
- It should instead go to `/success`, but doesn't.

To use the provided repository:
- `npm run ci`
- `npm run dev`
- go to "localhost:3000"
- Link should lead to /success, but instead goes to /failure

Happens in both dev and production builds.

## Expected behavior

Link should follow the redirect and go to `/success`, Best case, even show that redirect in the link preview beforehand.

## Additional context

This behavior did work as expected prior to next.js 9.5.3 (in 9.5.2 for example).

**Comment by ijjk** [comment]
Hi, the behavior you're describing does not work in previous versions of Next.js since `href` always maps to the page on the filesystem if it exists with `next/link`. If you want to redirect `/failure` to `/success` you should remove the `failure` file from your pages directory and use a normal anchor tag or map it manually with the link component using `href` and `as` e.g.

```jsx
<Link href='/success' as='failure'>
  <a>to /success as /failure</a>
</Link>
```

**Reply by lucasbasquerotto** [reply]
> ```js
> <Link href='/success' as='failure'>
>   <a>to /success as /failure</a>
> </Link>
> ```

This is not a viable solution in my case. I had conditions based on the presence of a cookie (`has: [{ type: 'cookie' ... }]`), and the redirect should not always happen (sometimes it should go to the page without redirecting). Is there a way to solve it in newer versions?

**If there isn't a solution yet, shouldn't a `bug` report issue be created?** I see the label `feature request`, but if redirects aren't redirecting, wouldn't it be a bug? Maybe in this case the redirect should be made in the page itself (in a `getStaticProps()` or `getServerSideProps()`), at least for now?

**Reply by ijjk** [reply]
This discussion is already a feature request and we're definitely keeping it in mind as handling these redirects automatically the same way we do for rewrites does seem desired. 

In the mean time you could use `getServerSideProps` to handle the mentioned redirect, correct.

**Reply by lucasbasquerotto** [reply]
In my project I'm not using `getServerSideProps` anywhere to save resources, and also because my main concern is SEO, so I just return static skeleton for pages, except for crawlers (and unauthenticated users in some pages).

What I do is including `getStaticProps` in the pages (for the skeletons and rendered pages), just changing a slug in the path to either: 1. fetch and render _or_ 2. return a skeleton (using rewrites so that the slug is based on a cookie value, without being actually in the path).

This issue caught me, tough. Fortunately, recently it was implemented redirects for `getStaticProps`, it worked but I received a lot of errors, the same as in the issue https://github.com/vercel/next.js/issues/20816, which was solved, although the stable version hasn't the fix yet, so I'm using the version `v11.0.2-canary.3` (as stated by this comment https://github.com/vercel/next.js/issues/20816#issuecomment-869841329).

This way I'm able now to redirect without including in the redirects array in `next.config.js`. Thanks!

**Reply by kendallstrautman** [reply]
@lucasbasquerotto I'm curious how you've implemented redirects via `getStaticProps`; I can't find documentation for this anywhere. When I look into the next.js types though I do see this:
```ts
export type GetStaticPropsResult<P> =
  | { props: P; revalidate?: number | boolean }
  | { redirect: Redirect; revalidate?: number | boolean }
  | { notFound: true }
```
So it seems like you can just check for the presence of the cookie in `getStaticProps` and then return the redirect object?

➕ to this feature request. It seems odd that we have the ability to set redirects based on the presence of a cookie in next.config but that wouldn't persist for links using `next/link`.

**Reply by lucasbasquerotto** [reply]
@kendallstrautman I define my page under a path like `/pages/_pages/[auth]/page/path` and use a rewrite (in `next.config.js`) that fills the `auth` parameter with a value indicating if the user is authenticated. In a page that is only accessed if the user is authenticated, I provide the correct value to the `auth` param (based on a cookie, using cookie matching, as explained at https://nextjs.org/docs/api-reference/next.config.js/rewrites#header-cookie-and-query-matching).

Then, in the page file, I do something like:

```tsx
export const getStaticProps: GetStaticProps = async ({ params }) => {
	if (params?.auth === 'true') {
		return {
			redirect: {
				destination: '/home',
				permanent: false,
			},
		};
	}

	const props = ...;
	return { props, revalidate: 3600 };
}
```

This is not exactly how I'm using it, but it follows this approach.

**Comment by vanandmkrtchyan** [comment]
If there were a way, a middlware like functionality BEFORE `getServersideProps`, that would be a great...

in my case I've wrapped my `getServersideProps` with HOF and checking if path exists in redirects list...

```javascript
export const preferRedirects = (callback) => (context) => {
  const { resolvedUrl } = context

  const path = resolvedUrl.split('?')[0]

  const redirect = redirects.find((item) => item.source === path)

  if (redirect) {
    return { redirect }
  }

  return callback(context)
}
```

**Comment by willemmulder** [comment]
Very annoying bug. I'd expect that if I set redirects in next.config.js, this would also be respected in the Link tags. 

Is this by any chance fixed in Next 13?

**Comment by bymoe** [comment]
What's the point of the Next Link ignoring the redirects put in the next.config.js? I just still do not understand the idea behind it.

**Comment by wottpal** [comment]
Did anyone find a workaround for this? Super frustrating behavior... Why do I define redirects in the first place 🤦‍♂️

**Reply by SimplySayHi** [reply]
As a - not perfect - workaround, I'm using `redirects` and `rewrites` together.
The redirect is not working when using NextJS Link  ( for example `<Link href="/">Link</Link>` ) component but on the homepage I can see the content of the page "/challenge".

```
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: false,
  swcMinify: true,
  async redirects() {
    return [
      {
        source: "/",
        destination: "/challenge",
        permanent: false,
      },
    ];
  },
  async rewrites() {
    return {
      beforeFiles: [
        {
          source: "/",
          destination: "/challenge",
        },
      ],
    };
  },
};

module.exports = nextConfig;
```

I would suggest NextJS team to test redirects also with next/link and router, since both are not working together with redirects from config 😊

**Comment by NabeelAhmed23** [comment]
Hi, Anyone found a solution? I am using Next 13 and facing this issue.

**Comment by bassem97** [comment]
yea, me too same problem here !!

**Reply by NabeelAhmed23** [reply]
I have found a way to do it. Don't know if it could be considered as a good solution but it works. Using a middleware to redirect

**Reply by bassem97** [reply]
how, can you show ?

**Reply by NabeelAhmed23** [reply]
I have created a middleware at the root level of the project,

import { NextResponse } from "next/server";
import redirections from "./redirects";

// eslint-disable-next-line consistent-return
export default function middleware(request) {
  // eslint-disable-next-line no-restricted-syntax, no-unreachable-loop
  for (const redirection of redirections) {
    if (request.nextUrl.pathname === redirection.source) {
      const url = new URL(request.nextUrl.origin + redirection.destination);
      return NextResponse.redirect(url, 308);
    }
  }
}

The redirections is an array with objects:
const redirections = [
  {
    source: {your source url},
    destination: {redirected url},
  },
.........
]

**Reply by bassem97** [reply]
i think this solution might work for you , just add this to next.config file inside module export 
```
    async redirects() {
        return [
            {
                source: '/old',
                destination: '/new',
                permanent: false,
            },
        ]
    },
    async rewrites() {
        return {
            beforeFiles: [
                {
                    source: '/old',
                    destination: '/new',
                },
            ],
        }
    },
```

**Reply by NabeelAhmed23** [reply]
I did try it, but it did not work if you go the page using next/link component. If i went to the page by putting the url in the address bar, then it would redirect me to the desired url but then if i used link to go the page, then it didnot redirected me to the desired url.

**Comment by timneutkens** [comment, marked as answer]
This limitation was kept in mind when building the new App Router and navigations now apply the redirects in middleware and next.config.js. For `pages` we can't do the same without introducing performance issues.

**Reply by leerob** [reply]
https://nextjs.org/docs/app/building-your-application/upgrading/app-router-migration

**Comment by kb1995** [comment]
I'm using v13 and the /pages directory, but I'm getting this behaviour. How can I opted out of it? 

There is definitely a bug here that needs to be fixed for the /pages directory, yes?

**Reply by kb1995** [reply]
@leerob's comment here https://github.com/vercel/next.js/issues/37889#issuecomment-1686875762 suggests that this has been fixed for the /app router but not for the /pages router.

This is currently impacting my production app...

**Reply by leerob** [reply]
Correct, this was not easily possible with the Pages Router, and we recommend either using `<a>` as a workaround here or incrementally adopting the App Router. It was one of the key things we wanted to solve with App Router.

**Reply by kb1995** [reply]
@leerob As far as I can see, this has never been acknowledged or documented by the Next.js team, outside of this Github discussion. 

The initial pitch of the App router was that everything with the Pages Router was going to remain stable for the indefinite future. Close to 1 year after the app router was released, we now can't use the <Link /> component within my pages folder.

**Reply by leerob** [reply]
The Pages Router is stable. This discussion doesn't mean that it's unstable, but it doesn't require you needing to update references in your codebase to the links that you're redirecting. For example, you add `/docs` to redirect to `/new` in `next.config.js`.

At the same time, you need to find and replace all instances of `/docs` in your codebase. These are the places where you've have client-side redirects with `next/link`, versus hitting the URL directly in the browser (hard nav, works as expected).

This is a bit annoying, and we wanted to solve this with the App Router. So that's why it was part of the design 👍
